include/seqan3/alphabet/detail/alphabet_base.hpp

```diff
@@ -0,0 +1,251 @@
+ *     the corresponding char value.
+ *   * `static std::array<rank_type, 256> constexpr char_to_rank` that defines for every possible character value the
+ *     corresponding rank value.
+ *
+ * ### Example
+ *
+ * This creates an alphabet called `ab` which has size two and the two letters 'A' and 'B':
+ * \snippet test/snippet/alphabet/detail/alphabet_base.cpp example
+ *
+ */
+template <typename derived_type, size_t size, typename char_t = char>
+class alphabet_base
+{
+public:
+    //TODO: this should be > 1 and there should be a different specialisation for size 1
+    static_assert(size > 0, "It does not make sense to use the base class for alphabets of size <= 1.");
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The type of the alphabet when converted to char (e.g. via to_char()).
+    using char_type = char_t;
+    //!\brief The type of the alphabet when represented as a number (e.g. via to_rank()).
+    using rank_type = detail::min_viable_uint_t<size - 1>;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_base() = default;
+    constexpr alphabet_base(alphabet_base const &) = default;
+    constexpr alphabet_base(alphabet_base &&) = default;
+    constexpr alphabet_base & operator=(alphabet_base const &) = default;
+    constexpr alphabet_base & operator=(alphabet_base &&) = default;
+    ~alphabet_base() = default;
+    //!\}
+
+    /*!\name Read functions
+     * \{
+     */
+    /*!\brief Return the letter as a character of char_type.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::to_char() requirement via the seqan3::to_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr char_type to_char() const noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
+    //!\endcond
+    {
+        return derived_type::rank_to_char[rank];
+    }
+
+    /*!\brief Return the letter's numeric value (rank in the alphabet).
+     *
+     * \details
+     *
+     * Satisfies the seqan3::semi_alphabet_concept::to_rank() requirement via the to_rank() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr rank_type to_rank() const noexcept
+    {
+        return rank;
+    }
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+    /*!\brief Assign from a character.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::assign_char() requirement via the seqan3::assign_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr derived_type & assign_char(std::conditional_t<std::Same<char_type, void>, char, char_type> const c) noexcept
```

Comment:
semi_alphabets don't have char types. So specialisations of alphabet_base  that are only semi-alphabet need to set this as void to prevent an overload of `assign_char` from being generated. Unfortunately even though the constraint would prevent the template from being instantiated when the type is void, the template would still be ill-formed, because void is not allowed for parameters. 

Therefore I need to wrap the type in the conditional.